jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/messages/ByteArrayMessageSink.java

```diff
@@ -79,38 +82,38 @@ public void accept(Frame frame, Callback callback)
                 if (out == null)
                     out = new ByteBufferCallbackAccumulator();
                 out.addEntry(payload, callback);
+                // The callback is now stored in the accumulator, so if
+                // the methodHandle throws, don't fail the callback twice.
+                callback = Callback.NOOP;
             }
 
-            // If the methodHandle throws we don't want to fail callback twice.
-            callback = Callback.NOOP;
             if (frame.isFin())
             {
                 byte[] buf = out.takeByteArray();
                 methodHandle.invoke(buf, 0, buf.length);
+                reset();
```

Comment:
Can we just make the `ByteBufferCallbackAccumulator` final?

I don't really understand why we need to be nulling it out and re-creating it for every message. Same deal for the other message sinks.

Reply:
I think the idea is to _not_ create it if the messages are always FIN single-frame.

However, if it gets created, we can keep it around, which is what I have done now.

The only glitch is for `StringMessageSink` where a large message will allocate a large `StringBuilder` and if then only small messages arrive, we are potentially wasting memory.
For binary sinks, the accumulator is a list that gets cleared so no problems with memory.
@gregw opinions?

Reply:
for binary sinks, I have a PR that merges the ByteBufferAccumulator and ByteBufferAggregator, so hopefully we can just retain the frames and avoid a copy until they have all arrived.

For string messages, I guess we could modify the Utf8StringBuilder so that if a small message is taken and the StringBuilder has a much larger capacity than was needed, then the reset nulls it rather than sets the length to 0.    Pity it doesn't have take symantics!

Reply:
Hmmm we could change the Utf8StringBuilder so that if the taken string was substantially smaller than the capacity, then we can do a StringBuilder.trimToSize after the reset.  Something like:
```java

    protected void bufferReset()
    {
        if ((_buffer.length() * 10) < _buffer.capacity())
            _buffer.trimToSize();
        _buffer.setLength(0);
    }
```

Reply:
Actually probably better to go to empty buffer and start again to determine new optimal length:
```java 
    protected void bufferReset()
    {
        boolean trim = _buffer.capacity() > (_buffer.length() * 8);
        _buffer.setLength(0);
        if (trim)
            _buffer.trimToSize();
    }
```

Reply:
> for binary sinks, I have a PR that merges the ByteBufferAccumulator and ByteBufferAggregator, so hopefully we can just retain the frames and avoid a copy until they have all arrived.

@gregw we already do this, the binary message sinks don't use `ByteBufferAccumulator` or `ByteBufferAggregator`. They use `ByteBufferCallbackAccumulator` which stores the buffers with their callbacks and only combines them when the full message has arrived.

Reply:
@lachlan-roberts see  #11058

Reply:
I've added this trim behavior to #11058